### PR TITLE
WPSC: Add Plugins tab

### PR DESCRIPTION
**Pull request by ockham**
Needs Automattic/wp-super-cache#382.

image

To test:
* On your WPSC test site, check out the `add/plugins_endpoint` branch.
* On your local Calypso dev installation, check out this branch, and edit `client/extensions/wp-super-cache/app/constants.js` to change `minVersion` to `'1.5.5'` for the `PLUGINS` tab (otherwise the 'Plugins' tab will be hidden.)
* Test the extension:
  * Navigate to `http://calypso.localhost:3000/extensions/wp-super-cache/<yourJPsite>`
  * Verify that the 'Plugins' tab is present, and that toggling the switches enables and disables the respective plugins (just verify that they're reflected by the corresponding `/wp-admin/options-general.php?page=wpsupercache&tab=plugins` screen).
* Change the 'Plugins' tab's `minVersion` back to `'1.5.6'`, and verify that the 'Plugins' tab is now hidden. Also verify that directly landing on `/extensions/wp-super-cache/plugins/<yourJPsite>` redirects you to `/plugins/wp-super-cache/<yourJPsite>`.

### CODE REVIEW

**Comment by donnapep** [comment]
It's a bit unclear what _Plugins_ means here. At first I assumed it meant the plugins that are activated on my site. I had to visit WP Admin to find out that they are WPSC-specific plugins. The Calypso _Plugins_ tab could benefit from some of that explanatory text that exists in WP Admin. And in turn I think that text could benefit from a copy review as I still had to read it twice to understand.

> On your local Calypso dev installation, check out this branch, and edit client/extensions/wp-super-cache/app/constants.js to change minVersion to '1.5.5' for the PLUGINS tab (otherwise the 'Plugins' tab will be hidden.)

This seems counter-intuitive to me. If the `/plugins` endpoint is only present in WPSC 1.5.6 or later, why is the _Plugins_ tab hidden when `minVersion` is 1.5.6 but visible when it's 1.5.5? Wouldn't it be the opposite?

> Also verify that directly landing on /extensions/wp-super-cache/plugins/<yourJPsite> redirects you to /plugins/wp-super-cache/<yourJPsite>.

I can confirm that it does that, but I'm wondering if it wouldn't be better to show them the _Easy_ tab instead. It's a bit jarring to be moved out of the settings altogether.

All of the toggles saved OK except for _No Adverts for Friends_. When I enabled that one, this is what WP Admin did:
adverts-for-friends

The spacing between the toggles and the explanation is off. It's closer to the following toggle than it is to the one it's associated with. I wonder if the external link indicator is necessary? Clicking through _Site Settings_ I found the _Content types_ box which has external links but no icons. Perhaps we could model these settings after those?
content-types

**Comment by ockham** [comment]
> A few thoughts after doing some testing. 😄 

> It's a bit unclear what _Plugins_ means here. At first I assumed it meant the plugins that are activated on my site. I had to visit WP Admin to find out that they are WPSC-specific plugins. The Calypso _Plugins_ tab could benefit from some of that explanatory text that exists in WP Admin. And in turn I think that text could benefit from a copy review as I still had to read it twice to understand.

Good point! Adding the current copy in 49a6669; pinging @michelleweber for review 😄

**Comment by ockham** [comment]
>> On your local Calypso dev installation, check out this branch, and edit client/extensions/wp-super-cache/app/constants.js to change minVersion to '1.5.5' for the PLUGINS tab (otherwise the 'Plugins' tab will be hidden.)

> This seems counter-intuitive to me. If the `/plugins` endpoint is only present in WPSC 1.5.6 or later, why is the _Plugins_ tab hidden when `minVersion` is 1.5.6 but visible when it's 1.5.5? Wouldn't it be the opposite?

No version containing the `/plugins` endpoint has been released yet; after talking to @donnchawp, I was just assuming it's going to be part of the next point release, i.e. 1.5.6. The WPSC `add/plugins_endpoint` branch however doesn't bump the version and is thus still at 1.5.5. So in order to have Calypso unhide the Plugins tab, we need to tell it that it's fine to use 1.5.5 for now :slightly_smiling_face:

**Comment by ockham** [comment]
>> Also verify that directly landing on /extensions/wp-super-cache/plugins/<yourJPsite> redirects you to /plugins/wp-super-cache/<yourJPsite>.

> I can confirm that it does that, but I'm wondering if it wouldn't be better to show them the _Easy_ tab instead. It's a bit jarring to be moved out of the settings altogether.

Indeed. Hmm, I'll have to modify `ExtensionRedirect` to allow a custom redirect target then... :thinking:

**Comment by ockham** [comment]
> All of the toggles saved OK except for _No Adverts for Friends_. When I enabled that one, this is what WP Admin did:
adverts-for-friends

Hmm, bummer. @donnchawp to the rescue?

**Comment by ockham** [comment]
> The spacing between the toggles and the explanation is off. It's closer to the following toggle than it is to the one it's associated with, and the spacing between multiple lines is too great. I wonder if the external link indicator is necessary? Clicking through _Site Settings_ I found the _Content types_ box which has external links but no icons. Perhaps we could model these settings after those?
content-types

Makes sense. Removed the icon per 79dcece.

**Comment by ockham** [comment]
>>> Also verify that directly landing on /extensions/wp-super-cache/plugins/<yourJPsite> redirects you to /plugins/wp-super-cache/<yourJPsite>.

>> I can confirm that it does that, but I'm wondering if it wouldn't be better to show them the _Easy_ tab instead. It's a bit jarring to be moved out of the settings altogether.

> Indeed. Hmm, I'll have to modify `ExtensionRedirect` to allow a custom redirect target then... :thinking:

Fixed per 3264479 and 414bc4e.

**Comment by ockham** [comment]
Thanks for your feedback @donnapep! I think I've addressed everything I could; waiting to hear from Editorial for copy, and for @donnchawp for https://github.com/Automattic/wp-calypso/pull/18002#issuecomment-333305992.

**Comment by donnapep** [comment]
> Indeed. Hmm, I'll have to modify ExtensionRedirect to allow a custom redirect target then...

I think this would actually be quite useful.

File: client/extensions/wp-super-cache/style.scss [hunk 1]
```diff
@@ -237,6 +237,12 @@
 	white-space: pre;
 }
 
+// Plugins Tab
+
+.wp-super-cache__plugins-tab .form-setting-explanation {
+	margin-left: 36px;
```
Comment (donnapep) [inline client/extensions/wp-super-cache/style.scss]: Could we also add 5px of bottom margin here? It's a wee bit too close to the next toggle option.

File: client/extensions/wp-super-cache/components/plugins/index.jsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	};
+
+	togglePlugin = ( plugin, newState ) => () => {
+		this.props.togglePlugin( this.props.siteId, plugin, newState );
+	}
+
+	render() {
+		const {
+			plugins,
+			isRequesting,
```
Comment (donnapep) [inline client/extensions/wp-super-cache/components/plugins/index.jsx]: Looks like `isRequesting` and `siteId` are missing from `propTypes`.

**Comment by ockham** [comment]
Thanks for reviewing @donnapep!

I'm going to merge this now, since it's been going for a while. Still waiting to hear from @Automattic/editorial about https://github.com/Automattic/wp-calypso/pull/18002#issuecomment-333305729, but I'm going to fix copy in a follow-up PR. (Due to the version guard, the Plugins tab isn't exposed to users yet anyway.)

**Comment by benhuberman** [comment]
@ockham for some reason Editorial pings aren't currently working properly (I saw this one via email) -- sorry for the delay, and we'll look into the issue.

As for the copy -- I'm not entirely clear on the context here; I agree this is very technical and "advanced"-sounding, though if it's only going to be showing in the Super Cache Plugins tab it's probably not as big a concern. If I misunderstood and this will be shown in a more "general audience" location, there might be room for additional tweaks (like explaining a little more about what these plugins/scripts/etc. are for). In the meantime, below is a suggested edit where I tried to make things a tad clearer / more easily parseable, but let me know if any of the technical aspects got warped in the process. Note: I changed "Supercache" in the first paragraph to "Super Cache" to keep it consistent with the plugin name, however if this instance refers to something else, feel free to revert to the previous spelling!

> Cache plugins are PHP scripts you'll find in a dedicated folder inside the WP Super Cache folder (wp-super-cache/plugins/). They load at the same time as WP Super Cache, and before regular WordPress plugins.
> 
> Keep in mind that cache plugins are for advanced users only. To create and manage them, you'll need extensive knowledge of both PHP and WordPress actions.
> 
> **Warning!** Due to the way WordPress upgrades plugins, the ones you upload to the WP Super Cache folder (wp-super-cache/plugins/) will be deleted when you upgrade WP Super Cache. To avoid this loss, load your cache plugins from a different location. When you set **$wp_cache_plugins_dir** to the new location in wp-config.php, WP Super Cache will look there instead. You can find additional details in the [developer documentation](https://odd.blog/wp-super-cache-developers/).

**Comment by ockham** [comment]
@benhuberman Thanks, love it! PR here: https://github.com/Automattic/wp-calypso/pull/18452

**Pull request merged**